File: core/src/main/java/javabot/database/UpgradeScript.java
package javabot.database;

import javabot.dao.ConfigDao;
import javabot.model.Config;

import javax.inject.Inject;
import java.io.IOException;
import java.sql.SQLException;
import java.util.Comparator;
import java.util.ServiceLoader;
import java.util.Set;
import java.util.TreeSet;

public abstract class UpgradeScript {
    @Inject
    private ConfigDao configDao;

    public final void execute() {
        final Config config = configDao.get();
        if (config.getSchemaVersion() < id()) {
            try {
                doUpgrade();
                registerUpgrade();
            } catch (Exception e) {
                throw new RuntimeException(e.getMessage(), e);
            }
        }
    }

    public void registerUpgrade() {
        final Config config = configDao.get();
        config.setSchemaVersion(id());
        configDao.save(config);
    }

    public abstract int id();

    public abstract void doUpgrade() throws SQLException, IOException;

    public final String toString() {
        return getClass().getName() + ":" + id();
    }
}

File: src/main/kotlin/javabot/database/UpgradeScript.kt
package javabot.database

import jakarta.inject.Inject
import java.io.IOException
import java.sql.SQLException
import javabot.dao.ConfigDao

abstract class UpgradeScript @Inject constructor(var configDao: ConfigDao) {
    fun execute() {
        val config = configDao.get()
        if (config.schemaVersion < id()) {
            try {
                doUpgrade()
                registerUpgrade()
            } catch (e: Exception) {
                throw RuntimeException(e.message, e)
            }
        }
    }

    fun registerUpgrade() {
        val config = configDao.get()
        config.schemaVersion = id()
        configDao.save(config)
    }

    abstract fun id(): Int

    @Throws(SQLException::class, IOException::class) abstract fun doUpgrade()

    override fun toString(): String {
        return javaClass.name + ":" + id()
    }
}
